packages/styled-components/src/models/StyledNativeComponent.js

```diff
@@ -1,152 +1,148 @@
-    const { inlineStyle } = props.forwardedComponent;
+  // NOTE: the non-hooks version only subscribes to this when !componentStyle.isStatic,
+  // but that'd be against the rules-of-hooks. We could be naughty and do it anyway as it
+  // should be an immutable value, but behave for now.
+  const theme = determineTheme(props, useContext(ThemeContext), defaultProps);
 
-    const executionContext = this.buildExecutionContext(
-      theme,
-      props,
-      props.forwardedComponent.attrs
-    );
+  const [context, attrs] = useResolvedAttrs(theme || EMPTY_OBJECT, props, componentAttrs);
 
-    return inlineStyle.generateStyleObject(executionContext);
-  }
+  const generatedStyles = inlineStyle.generateStyleObject(context);
 
-  setNativeProps(nativeProps: Object) {
-    if (this.root !== undefined) {
-      // $FlowFixMe
-      this.root.setNativeProps(nativeProps);
-    } else if (process.env.NODE_ENV !== 'production') {
-      // eslint-disable-next-line no-console
-      console.warn(
-        'setNativeProps was called on a Styled Component wrapping a stateless functional component.'
-      );
+  const refToForward = forwardedRef;
+
+  const elementToBeCreated: Target = attrs.$as || props.$as || attrs.as || props.as || target;
+
+  const isTargetTag = isTag(elementToBeCreated);
+  const computedProps = attrs !== props ? { ...props, ...attrs } : props;
+  const propFilterFn = shouldForwardProp || (isTargetTag && validAttr);
+  const propsForElement = {};
+
+  // eslint-disable-next-line guard-for-in
+  for (const key in computedProps) {
+    if (key[0] === '$' || key === 'as') continue;
+    else if (key === 'forwardedAs') {
+      propsForElement.as = computedProps[key];
+    } else if (!propFilterFn || propFilterFn(key, validAttr)) {
+      // Don't pass through non HTML tags through to HTML elements
+      propsForElement[key] = computedProps[key];
     }
   }
+
+  propsForElement.style = [generatedStyles].concat(props.style || []);
+
+  propsForElement.ref = refToForward;
+
+  return createElement(elementToBeCreated, propsForElement);
 }
 
 export default (InlineStyle: Function) => {
   const createStyledNativeComponent = (target: Target, options: Object, rules: RuleSet) => {
+    const isTargetStyledComp = isStyledComponent(target);
+    const isCompositeComponent = !isTag(target);
+
     const {
-      attrs = EMPTY_ARRAY,
       displayName = generateDisplayName(target),
-      ParentComponent = StyledNativeComponent,
+      componentId = generateId(options.displayName, options.parentComponentId),
+      attrs = EMPTY_ARRAY,
     } = options;
 
-    const isClass = !isTag(target);
-    const isTargetStyledComp = isStyledComponent(target);
-
-    // $FlowFixMe
-    const WrappedStyledNativeComponent = React.forwardRef((props, ref) => (
-      <ParentComponent
-        {...props}
-        forwardedComponent={WrappedStyledNativeComponent}
-        forwardedRef={ref}
-      />
-    ));
+    const styledComponentId =
```

Comment:
I’m not sure if this is needed for native, it seems to be only used by useDebugValue.

Reply:
Yeah this isn't needed, we don't support component selectors in native because of the lack of cascade